### PR TITLE
fixing _process_result

### PR DESCRIPTION
**Pull request by mkoura**
The response to action performed on resource(s) can contain either data corresponding to collection where the action took place, or to other collection like "service_requests", "automation_requests", etc.

We always expect that the "href" contained in the response corresponds to the collection where the action took place, which is incorrect. In case of e.g. ``service_template.action.order()``, the returned entity belongs to ``/api/collections/service_catalogs/:id/service_templates``collection instead of ``/api/service_requests`` collection where it should belong to.

The fix: select correct collection based on the "href" contained in the response.

### CODE REVIEW

File: src/manageiq_client/api.py [hunk 1]
```diff
@@ -616,11 +616,33 @@ def __call__(self, *args, **kwargs):
...
+            entry_point = href_match.group(1)
+            if collection_name and entry_point:
+                new_collection = Collection(
+                    self.collection.api,
+                    "{}/{}".format(entry_point, collection_name),
+                    collection_name
+                )
+                return Entity(new_collection, result, incomplete=True)
+
+        return Entity(self.collection, result, incomplete=True)
```
Comment (cben) [inline src/manageiq_client/api.py]: [cosmetic] consider making this the early return, so the complex path will not be inside conditionals.

Hmm, as you wrote it, when `if href_match` or `if collection_name and entry_point` checks fail, we fall through to this return from `self.collection`.  Is that the right behavior, or should it raise an exception?

File: src/manageiq_client/api.py [hunk 2]
```diff
@@ -616,11 +616,33 @@ def __call__(self, *args, **kwargs):
...
+        if self.collection._href not in href_result:
+            href_match = re.match(r"(http[^?]+)/([a-z_-]+)", href_result)
+
+        if href_match:
+            collection_name = href_match.group(2)
+            entry_point = href_match.group(1)
+            if collection_name and entry_point:
+                new_collection = Collection(
+                    self.collection.api,
+                    "{}/{}".format(entry_point, collection_name),
```
Comment (cben) [inline src/manageiq_client/api.py]: this essentially constructs back `href_result`, right?
do you need the RE matching & splitting then?  oh, I see, for the collection_name. :+1:

File: testing/test_process_result.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+from manageiq_client.api import Collection, ActionContainer, Action
+
+
+def test_get_entity_from_href():
+    url_base = "http://example.com/api/"
+    test_data = [
```
Comment (cben) [inline testing/test_process_result.py]: can use parametric test (https://docs.pytest.org/en/latest/parametrize.html) for clearer error attribution if one of these fails.

**Review by cben — COMMENTED** [review]
Ooh, nice!
Does this cover what the special case code that #25 gets rid of was trying to cover?

One question about error handling, otherwise LGTM.

File: src/manageiq_client/api.py (hunk 1 above)
Comment (mkoura) [reply src/manageiq_client/api.py]: The intent was to be backwards compatible and fall through if the attempt to find "better" collection failed. Even in wrong collection the data can still be usable, because the href points to an existing location (it can be reloaded, get up-to-date message etc.).

**Review by cben — APPROVED** [review]

File: src/manageiq_client/api.py [hunk 4]
```diff
@@ -616,11 +616,33 @@ def __call__(self, *args, **kwargs):
             self.api.response = action_response
         return outcome
 
+    def _get_entity_from_href(self, result):
+        href_result = result['href']
+        href_match = None
+
+        # If the "href" value in result doesn't match the current collection,
+        # try to find the collection that the "href" refers to.
+        if self.collection._href not in href_result:
```
Comment (simon3z) [inline src/manageiq_client/api.py]: @mkoura do you think this can be a little more strict? e.g. something like `not self.collection._href.startswith(href_result)`

File: src/manageiq_client/api.py [hunk 5]
```diff
@@ -616,11 +616,33 @@ def __call__(self, *args, **kwargs):
...
+            href_match = re.match(r"(http[^?]+)/([a-z_-]+)", href_result)
+
+        if href_match:
```
Comment (simon3z) [inline src/manageiq_client/api.py]: @mkoura I know not everyone agrees with multiple function exits but I think that it would be much simpler if we reverse the logic a little bit:

```python
def _get_entity_from_href(self, result):
    result_href = result['href']

    if self.collection._href.startswith(result_href):
        return Entity(self.collection, result, incomplete=True)

    ...
```

so we would spare this extra condition.

File: src/manageiq_client/api.py (hunk 4 above)
Comment (mkoura) [reply src/manageiq_client/api.py]: Agree, this is better.

File: src/manageiq_client/api.py (hunk 5 above)
Comment (mkoura) [reply src/manageiq_client/api.py]: Ok, I don't have problem with that.

File: src/manageiq_client/api.py [hunk 6]
```diff
@@ -617,26 +617,27 @@ def __call__(self, *args, **kwargs):
...
-                    "{}/{}".format(entry_point, collection_name),
-                    collection_name
-                )
-                return Entity(new_collection, result, incomplete=True)
 
-        return Entity(self.collection, result, incomplete=True)
+        if self.collection._href.startswith(href_result):
+            return Entity(self.collection, result, incomplete=True)
+
+        href_match = re.match(r"(https?://.+/api[^?]*)/([a-z_-]+)", href_result)
```
Comment (mkoura) [inline src/manageiq_client/api.py]: @simon3z @cben To be sure that the data obtained from regex are meaningful (not that it just matched something) I changed this to more complex regex and raise exception if it doesn't match. This way it was also possible to simplify the rest of the logic.

I added more tests to make sure this works as expected in all corner cases.

File: testing/test_process_result.py [hunk 7]
```diff
@@ -4,6 +4,12 @@
 from manageiq_client.api import Collection, ActionContainer, Action
 
 
+url_bases = [
+    "https://example.com/api/v2.4.0/",
+    "https://example.com/api/",
+    "https://127.0.0.1:8080//api",
```
Comment (simon3z) [inline testing/test_process_result.py]: @mkoura do we support also plain `http`? Do you want to add a test to make sure it can be used?

Comment (mkoura) [reply testing/test_process_result.py]: Sure, will add.

File: testing/test_process_result.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
+# -*- coding: utf-8 -*-
+import pytest
+
+from manageiq_client.api import Collection, ActionContainer, Action
+
+
+url_bases = [
+    "https://example.com/api/v2.4.0/",
+    "https://example.com/api/",
+    "https://127.0.0.1:8080//api",
```
Comment (simon3z) [inline testing/test_process_result.py]: @mkoura do we support also plain http? Do you want to add a test to make sure it can be used?
